### PR TITLE
[SPARK-12526][SPARKR]`ifelse`, `when`, `otherwise` unable to take Column as value

### PR DESCRIPTION
**Pull request by saurfang**
`ifelse`, `when`, `otherwise` is unable to take `Column` typed S4 object as values.

For example:

``` r
ifelse(lit(1) == lit(1), lit(2), lit(3))
ifelse(df$mpg > 0, df$mpg, 0)
```

will both fail with

``` r
attempt to replicate an object of type 'environment'
```

The PR replaces `ifelse` calls with `if ... else ...` inside the function implementations to avoid attempt to vectorize(i.e. `rep()`). It remains to be discussed whether we should instead support vectorization in these functions for consistency because `ifelse` in base R is vectorized but I cannot foresee any scenarios these functions will want to be vectorized in SparkR.

For reference, added test cases which trigger failures:

``` r
. Error: when(), otherwise() and ifelse() with column on a DataFrame ----------
error in evaluating the argument 'x' in selecting a method for function 'collect': 
  error in evaluating the argument 'col' in selecting a method for function 'select': 
  attempt to replicate an object of type 'environment'
Calls: when -> when -> ifelse -> ifelse

1: withCallingHandlers(eval(code, new_test_environment), error = capture_calls, message = function(c) invokeRestart("muffleMessage"))
2: eval(code, new_test_environment)
3: eval(expr, envir, enclos)
4: expect_equal(collect(select(df, when(df$a > 1 & df$b > 2, lit(1))))[, 1], c(NA, 1)) at test_sparkSQL.R:1126
5: expect_that(object, equals(expected, label = expected.label, ...), info = info, label = label)
6: condition(object)
7: compare(actual, expected, ...)
8: collect(select(df, when(df$a > 1 & df$b > 2, lit(1))))
Error: Test failures
Execution halted
```

### CODE REVIEW

**Comment by shivaram** [comment]
Thanks @saurfang for the PR. I think this is a good enough fix as this is in an internal implementation detail and not visible to SparkR users ? 

cc @sun-rui

**Comment by saurfang** [comment]
Indeed. Like I said I don't see compelling reason to use these three functions in vectorized way. Let me know if you have any other comments on the fix.

**Comment by sun-rui** [comment]
The fix is good, but some style nit:
if (...) { ... } else { ... }

File: R/pkg/R/column.R [hunk 1]
```diff
@@ -225,7 +225,7 @@ setMethod("%in%",
 setMethod("otherwise",
           signature(x = "Column", value = "ANY"),
           function(x, value) {
-            value <- ifelse(class(value) == "Column", value@jc, value)
+            value <- if(class(value) == "Column") { value@jc } else { value }
```
Comment (sun-rui) [inline R/pkg/R/column.R]: if<space>( :)

**Comment by saurfang** [comment]
Thanks for the review @sun-rui. Hope that's better. Looks like `lintr`, as awesome as it is, let that slip through, which I have filed a separate issue here: https://github.com/jimhester/lintr/pull/128

**Comment by shivaram** [comment]
LGTM. Thanks @saurfang and @sun-rui - Merging this

**Pull request closed without merging**